src/Http/Http.Extensions/gen/RequestDelegateGenerator.cs

```diff
@@ -39,83 +41,141 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
-                        }
-                        {{StaticRouteHandlerModelEmitter.EmitFilteredInvocation()}}
-                    },
-                    options.EndpointBuilder,
-                    handler.Method);
-                }
+                            if (ic.HttpContext.Response.StatusCode == 400)
+                            {
+                                return ValueTask.FromResult<object?>(Results.Empty);
+                            }
+{{endpoint.EmitFilteredInvocation()}}
+                        },
+                        options.EndpointBuilder,
+                        handler.Method);
+                    }
 
-                {{StaticRouteHandlerModelEmitter.EmitRequestHandler()}}
-                {{StaticRouteHandlerModelEmitter.EmitFilteredRequestHandler()}}
+{{endpoint.EmitRequestHandler()}}
+{{StaticRouteHandlerModelEmitter.EmitFilteredRequestHandler()}}
 
-                RequestDelegate targetDelegate = filteredInvocation is null ? RequestHandler : RequestHandlerFiltered;
-                var metadata = inferredMetadataResult?.EndpointMetadata ?? ReadOnlyCollection<object>.Empty;
-                return new RequestDelegateResult(targetDelegate, metadata);
-            }),
+                    RequestDelegate targetDelegate = filteredInvocation is null ? RequestHandler : RequestHandlerFiltered;
+                    var metadata = inferredMetadataResult?.EndpointMetadata ?? ReadOnlyCollection<object>.Empty;
+                    return new RequestDelegateResult(targetDelegate, metadata);
+                }),
 """);
 
-        var stronglyTypedEndpointDefinitions = endpoints.Select((endpoint, _) => $$"""
-        internal static global::Microsoft.AspNetCore.Builder.RouteHandlerBuilder {{endpoint.HttpMethod}}(
-            this global::Microsoft.AspNetCore.Routing.IEndpointRouteBuilder endpoints,
-            [global::System.Diagnostics.CodeAnalysis.StringSyntax("Route")] string pattern,
-            global::{{StaticRouteHandlerModelEmitter.EmitHandlerDelegateType(endpoint)}} handler,
-            [global::System.Runtime.CompilerServices.CallerFilePath] string filePath = "",
-            [global::System.Runtime.CompilerServices.CallerLineNumber]int lineNumber = 0)
-        {
-            return global::Microsoft.AspNetCore.Http.Generated.GeneratedRouteBuilderExtensionsCore.MapCore(endpoints, pattern, handler, {{StaticRouteHandlerModelEmitter.EmitVerb(endpoint)}}, filePath, lineNumber);
-        }
-""");
+        var stronglyTypedEndpointDefinitions = endpoints
+            .Collect()
+            .Select((endpoints, _) =>
+            {
+                var dedupedByDelegate = endpoints.Distinct(new LambdaComparer<Endpoint>((a, b) =>
+                {
+                    if (a.Response.IsAwaitable == b.Response.IsAwaitable &&
+                        a.Response.IsVoid == b.Response.IsVoid &&
+                        SymbolEqualityComparer.Default.Equals(a.Response.ResponseType, b.Response.ResponseType) &&
```

Comment:
```suggestion
                    if (a.Response.WrappedResponseType == b.Response.WrappedResponseType && 
```

The call `EmitHandlerDelegateType()` uses the `WrappedResponseType`, so we need to compare those. Currently, the test case below only produces a MapGet overload for `Func<Task<string>>` and not `Func<ValueTask<string>>`.

```csharp
    [Fact]
    public async Task Multiple_MapAction_NoParam_TaskAndValueTaskReturn()
    {
        var source = """
app.MapGet("/en", () => Task.FromResult("Hello world!"));
app.MapGet("/es", () => new ValueTask<string>("Hola mundo!"));
""";
        var (_, compilation) = RunGenerator(source);

        await VerifyAgainstBaselineUsingFile(compilation);
    }
```

Reply:
Good scenario! I added it to the tests.